README.md
<div align="center">

[![Release](https://img.shields.io/github/v/release/wayfair/one-version?display_name=tag)](CHANGELOG.md)
[![license: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](LICENSE)
[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-2.0-4baaaa.svg)](CODE_OF_CONDUCT.md)
[![Maintainer](https://img.shields.io/badge/Maintainer-Wayfair-7F187F)](https://wayfair.github.io)

</div>

<h2 align="center">@wayfair/one-version</h2>
<div align="center" >
<i>One Version to rule them all, One Version to find them,

One Version to bring them all, and in the darkness bind them.<sup>1</sup>
</i>

</div>

<h3 align="center">Opinionated Monorepo Dependency Management CLI</h3>

**🚨 Enforcement**: Require all workspaces in a monorepo to conform to the [One-Version rule](#one-version-rule).

**📦 Supports multiple package managers**: Support for `yarn` classic, `yarn` berry, and `pnpm` workspaces.

**💥 Coordinated upgrades**: Coming Soon!

---

## Table Of Contents

- [Table Of Contents](#table-of-contents)
- [One-Version Rule](#one-version-rule)
- [Getting Started](#getting-started)
- [Configuration](#configuration)
  - [Supported Options](#supported-options)
    - [overrides (optional, object)](#overrides-optional-object)
  - [Examples](#examples)
- [CLI](#cli)
  - [`-f`, `--file` (optional, string)](#-f---file-optional-string)
- [Contributing](#contributing)
- [License](#license)

## One-Version Rule

This package implements a version of Google's `One-Version Rule`:

> For every dependency in [a] repository, there must be only one version of that dependency to choose.<sup>2</sup>

Please refer to the [implementation notes](ONE-VERSION.md) for our specific evaluation criteria.

There is some overlap between this tool and [experimental yarn constraints](https://yarnpkg.com/features/constraints), without requiring use of a particular package manager.

## Getting Started

Install `@wayfair/one-version` at the workspace root using yarn:

```bash
yarn add --dev -w @wayfair/one-version
```

Or pnpm:

```bash
pnpm add -save-dev -w @wayfair/one-version
```

Add the following section to your package.json:

```json
{
  "scripts": {
    "one-version:check": "one-version check"
  }
}

```

Run `yarn one-version:check` or `pnpm run one-version:check`.

If the repo is compliant, the tool will print this message:

```text
✨ One Version Rule Success - found no version conflicts!
```  

If the repo is not compliant, you will see a version of this message:

```text
🚫 One Version Rule Failure - found multiple versions of the following dependencies:

prettier
  2.1.2
    dev: @wayfair/app-a, @wayfair/app-b
  ^2.3.2
    dev: @wayfair/app-c
  2.2.1
    dev: @wayfair/lib-a
```

## Configuration

The behavior of `@wayfair/one-version` can be configured by a `one-version.config.json` at the root of the repository.

### Supported Options

#### overrides (optional, object)

Overrides lets workspaces opt out of the one-version rule. This may be useful while performing major upgrades.

### Examples

```json
{
  "overrides": {
    "dependency": {
      "versionSpecifier": ["workspaceA", "workspaceB"]
    }
  }
}
```

For example, the below config will allow `app-A` and `lib-L` to specify `react@^16.9`, even if the rest of the repo specifies `react@^17`.

```json
{
  "overrides": {
    "react": {
      "^16.9": ["app-A", "lib-L"]
    }
  }
}
```

## CLI

### `-f`, `--file` (optional, string)

Pass a path to a local workspace to include in the one version check. This is useful for preparing a workspace to be migrated into a monorepo that follows the one version rule.

Example: `yarn one-version:check -f ../path/to/local/workspace`

## Contributing

Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any contributions you make are **greatly appreciated** 💜. For contributing guidelines, please see [CONTRIBUTING.md](CONTRIBUTING.md)

## License

Distributed under the `MIT` License. See `LICENSE` for more information.

---

`1`: J.R.R. Tolkien, 1954. Mostly.

`2`: [Software Engineering At Google](https://abseil.io/resources/swe_at_google.2.pdf) - Winters, Manshreck and Wright, 2020, p. 341


CHANGELOG.md
# Changelog

This project adheres to [semver](https://semver.org/spec/v2.0.0.html).
This changelog format is largely based on [Keep A Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased]

#### 💥 Breaking Changes

#### 🚀 New Features & Enhancements

#### 🐛 Bug Fixes

#### 🔨 Chores & Documentation

## [0.5.0] - 2023-06-05

#### 💥 Breaking Changes

- Drop support for Node 14 (#43)

## [0.4.0] - 2023-02-15

#### 🚀 New Features & Enhancements

- Support a -f flag for arbitrary file inclusion in the check

#### 🔨 Chores & Documentation

- Refactored check api to unify across package managers
- Use modern Yarn internally

## [0.3.0] - 2022-06-27

#### 🚀 New Features & Enhancements

- Add detection and support for Yarn Berry

## [0.2.0] - 2022-05-16

#### 🚀 New Features & Enhancements

- Detect package manager based on lockfile

## [0.1.3] - 2022-05-03

#### 🔨 Chores & Documentation

- Remove fixtures dir from published package

## [0.1.2] - 2022-05-01

#### 🔨 Chores & Documentation

- Add `files` config to pjson, removes unnecessary files from published package
- Update node `engines` config to allow for `^16` and `^18`

## [0.1.1] - 2022-05-01

#### 🐛 Bug Fixes

- Fix bug in cli program name after package rename
- Republish with new public repo settings

## [0.1.0] - 2022-04-29

#### 🚀 New Features & Enhancements

- Initial package. Full yarn support.

## [0.0.0] - 2022-04-29

#### 🔨 Chores & Documentation

- Initial repo setup, empty release


src/shared/util.js
const { readFileSync, existsSync } = require("fs");
const path = require("path");
const {
  DEPENDENCY_TYPES,
  CONFIG_FILE,
  YARN_RC,
  YARN_LOCK,
  PNPM_LOCK,
} = require("./constants");

/**
 * Parse a config file if it exists
 */
const parseConfig = (configFilePath = CONFIG_FILE) => {
  const configPath = path.join(process.cwd(), configFilePath);
  const configContents = existsSync(configPath)
    ? readFileSync(configPath, { encoding: "utf-8" })
    : "{}";
  return JSON.parse(configContents);
};

/**
 * Read the manifest at a specified path, return only the fields we care about
 */
const getPackageDeps = (packageRoot) => {
  const packageContents = readFileSync(path.join(packageRoot, "package.json"), {
    encoding: "utf8",
  });
  const { name, peerDependencies, devDependencies, dependencies } =
    JSON.parse(packageContents);
  return { name, peerDependencies, devDependencies, dependencies };
};

/**
 * Creates or updates the version dependencies for a given package
 */
const addOrUpdateVersion = ({ seenVersions, type, version, consumerName }) => {
  const seenConsumers = seenVersions?.[version]?.[type] || [];
  const versionConsumers = seenConsumers.concat(consumerName);
  return {
    ...seenVersions,
    [version]: {
      ...seenVersions?.[version],
      [type]: versionConsumers,
    },
  };
};

/**
Transform dependencies from an array of package json formats, i.e. multiple
{
  name: "packageName",
  dependencies: {
    'react': '^18'
  }
}
into an inverted structure organized by each dependency name, version, and type:
{
  react: {
    '^18': {direct: [ 'packageName' ], peer: ['demo-package']},
    '^17': {direct: [ 'demo', 'platform-capabilities' ]}
  },
}
that is,
{
   dependencyName: {
     versionSpecifier: {dependencyType: [ consumers using this specifier and type ]}
   }
}
*/
const transformDependencies = (manifests) => {
  return manifests.reduce(
    (
      acc,
      { name: consumerName, dependencies, peerDependencies, devDependencies }
    ) => {
      if (dependencies) {
        Object.entries(dependencies).forEach(([packageName, version]) => {
          acc[packageName] = addOrUpdateVersion({
            seenVersions: acc[packageName],
            type: DEPENDENCY_TYPES.DIRECT,
            version,
            consumerName,
          });
        });
      }
      if (peerDependencies) {
        Object.entries(peerDependencies).forEach(([packageName, version]) => {
          acc[packageName] = addOrUpdateVersion({
            seenVersions: acc[packageName],
            type: DEPENDENCY_TYPES.PEER,
            version,
            consumerName,
          });
        });
      }
      if (devDependencies) {
        Object.entries(devDependencies).forEach(([packageName, version]) => {
          acc[packageName] = addOrUpdateVersion({
            seenVersions: acc[packageName],
            type: DEPENDENCY_TYPES.DEV,
            version,
            consumerName,
          });
        });
      }
      return acc;
    },
    {}
  );
};

/**
 * Removes overridden dependencies from the versions arrays
 */
const removeOverriddenDependencies = ({ packageOverrides, versions }) => {
  return Object.entries(versions)
    .map(([version, { direct, peer, dev }]) => {
      const filteredPackages = {};
      const notOverridden = (packageName) =>
        !packageOverrides[version]?.includes(packageName);
      if (direct) {
        const directDependencies = direct.filter(notOverridden);
        if (directDependencies.length > 0) {
          filteredPackages[DEPENDENCY_TYPES.DIRECT] = directDependencies;
        }
      }
      if (peer) {
        const peerDependencies = peer.filter(notOverridden);
        if (peerDependencies.length > 0) {
          filteredPackages[DEPENDENCY_TYPES.PEER] = peerDependencies;
        }
      }
      if (dev) {
        const devDependencies = dev.filter(notOverridden);
        if (devDependencies.length > 0) {
          filteredPackages[DEPENDENCY_TYPES.DEV] = devDependencies;
        }
      }
      return [version, filteredPackages];
    })
    .filter(([, dependents]) => Object.keys(dependents).length > 0);
};

/**
 * Finds dependencies with multiple versions (excluding overrides)
 */
const findDuplicateDependencies = (dependencies, overrides) => {
  return Object.entries(dependencies)
    .map(([packageName, versions]) => {
      const packageOverrides = overrides?.[packageName];
      if (packageOverrides) {
        const filteredVersions = removeOverriddenDependencies({
          packageOverrides,
          versions,
        });
        return [packageName, Object.fromEntries(filteredVersions)];
      }
      return [packageName, versions];
    })
    .filter(([, versions]) => Object.keys(versions).length > 1);
};

/**
 * Detect the package manager being used by the project
 */
const detectPackageManager = () => {
  if (existsSync(YARN_LOCK)) {
    if (existsSync(YARN_RC)) {
      return "berry";
    }
    return "yarn";
  }
  if (existsSync(PNPM_LOCK)) {
    return "pnpm";
  }
  return "";
};

/**
 * Validate that a path exists
 */
const isValidPath = (filePath) => {
  if (existsSync(path.join(filePath, "package.json"))) {
    return true;
  } else {
    throw new Error(`Invalid workspace: ${filePath}`);
  }
};

module.exports = {
  parseConfig,
  getPackageDeps,
  transformDependencies,
  findDuplicateDependencies,
  detectPackageManager,
  isValidPath,
};
